### ISSUE TITLE
DroneLink - Recieve and send communication to the drone

### ISSUE THREAD

**Issue by Crowdedlight** [DronesForBlood/DronesForBlood#15]
- [x] Receive pos (already gets published, just grab those messages)
- [x] Upload full mission
- [x] Add retries functionalities to mission_upload
- [ ] Figure out why "wp index out of bounds" error gets thrown
- [x] Send start mission
- [x] Implement Command protocol with retries and timeouts
- [x] Implement command ack topic. Respond with both cmd_id and result as string
- [ ] Interface download mission
- [x] Implement heartbeat message
- [x] Arm/disarm interface and command
- [x] Create ack topic for mission ack, to respond accordingly.  
- [x] Clear mission interface (Before gaining a new mission)
- [x] Implement command MAV_CMD_DO_SET_MODE
- [x] Go to given pos (might be needed when updating mission list)
~~- [ ] Get next waypoint and endpoint~~
~~- [ ] Sending commands like stop, stay pos, continue mission~~
- [x] Implement MAV_CMD_NAV_GUIDED_ENABLE interface, to enable setting flight-mode as offbord. This is needed if wanting to use guided method, to avoid stops of drone, when changing path
- [x] implement guided method, of sending single waypoints when we get close to current waypoints, to remove stops when updating route
~~- [ ] Implement interface GPS_GLOBAL_ORIGIN \#49 so 0,0,0 origin can be recieved for proper navigation~~
- [x] implement precision land interface

**Comment by Crowdedlight** [comment]
On Hold, waiting on Mavlink_lora module.

**Comment by Crowdedlight** [comment]
Extended mavlink lora module. Can now upload missions. Waiting on feedback from Kjeld before extending further with interface. Added list of cmds to implement interface for.

**Comment by Crowdedlight** [comment]
Heartbeat implemented with 1hz output. Set timeout on pixhawk accordingly.

**Comment by Crowdedlight** [comment]
Redo heartbeat, so lora just relays GCS heartbeat. So drone knows that if GCS_master fails, or isn't active, the heartbeat stops even if Lora is running

**Comment by Crowdedlight** [comment]
GPS_GLOBAL_ORIGIN isn't supported by px4 currently.

**Comment by Crowdedlight** [comment]
Implement Following interfaces with high priority

- [ ] Mission download
- [x] MISSION_CURRENT
~~- [ ] MISSION_ITEM_REACHED <-- Does this output on px4 for each waypoint, or just for mission end?~~
- [x] MAV_CMD_DO_REPOSITION <-- For manual positioning if needed

**Comment by Crowdedlight** [comment]
MISSION_ITEM_REACHED <-- Only outputs at end of mission

**Comment by Crowdedlight** [comment]
DO_CHANGE_HEIGHT

**Comment by Crowdedlight** [comment]
Do change height and nav follow and change position doesn't seem to be implemented in px4 stack.

**Comment by jlrandulfe** [comment]
Is this issue over? If there are new requirements/bugs regarding the dronelink, maybe they can be created in new issues

**Comment by Crowdedlight** [comment]
Only think not done so-far we might want is mission download. However not sure if we plan on doing that. Otherwise we can create a new issue with that.
